expertsystem/data.py

```diff
@@ -0,0 +1,191 @@
+
+    def __repr__(self) -> str:
+        return (
+            f'<{self.__class__.__name__} {"+1" if self.__value > 0 else "-1"}>'
+        )
+
+    @property
+    def value(self) -> int:
+        return self.__value
+
+
+class Spin:
+    """Safe, immutable data container for (iso)spin."""
+
+    def __init__(self, magnitude: float, projection: float) -> None:
+        if abs(projection) > magnitude:
+            raise ValueError(
+                "Spin projection cannot be larger than its magnitude"
+            )
+        self.__magnitude = float(magnitude)
+        self.__projection = float(projection)
+
+    def __eq__(self, other: object) -> bool:
+        return self.__magnitude == other
+
+    def __float__(self) -> float:
+        return self.__magnitude
+
+    def __repr__(self) -> str:
+        return f"<{self.__class__.__name__} {self.__magnitude, self.__projection}>"
+
+    @property
+    def magnitude(self) -> float:
+        return self.__magnitude
+
+    @property
+    def projection(self) -> float:
+        return self.__projection
+
+
+class MeasuredValue(NamedTuple):
+    """Value with (optional) uncertainty, as reported by a measurement."""
+
+    value: float
+    uncertainty: Optional[float] = None
+
+    def __eq__(self, other: object) -> bool:
+        return self.value == other
+
+    def __float__(self) -> float:
+        return self.value
+
+    def __repr__(self) -> str:
+        if self.uncertainty is None:
+            return str(self.value)
+        return f"{self.value} ± {self.uncertainty}"
+
+
+class Particle(NamedTuple):
+    """Immutable data container for particle info."""
+
+    name: str
+    pid: int
+    charge: float
+    spin: float
+    mass: MeasuredValue
+    strange: int = 0
+    charm: int = 0
+    bottom: int = 0
+    top: int = 0
+    baryon: int = 0
+    ln_e: int = 0
+    ln_mu: int = 0
+    ln_tau: int = 0
+    width: Optional[MeasuredValue] = None
+    isospin: Optional[Spin] = None
+    parity: Optional[Parity] = None
+    cparity: Optional[Parity] = None
+    gparity: Optional[Parity] = None
+
+
+class ParticleCollection:
+    """Safe, `dict`-like collection of `.Particle` instances."""
+
+    def __init__(self) -> None:
+        self.__particles: Dict[str, Particle] = dict()
+
+    def __getitem__(self, particle_name: str) -> Particle:
+        return self.__particles[particle_name]
+
+    def __contains__(self, particle_name: str) -> bool:
+        return particle_name in self.__particles
+
+    def __len__(self) -> int:
+        return len(self.__particles)
+
+    def __repr__(self) -> str:
+        return str(self.__particles)
+
+    def add(self, particle: Particle) -> None:
+        self.__particles[particle.name] = particle
+
+    def items(self) -> ItemsView[str, Particle]:
+        return self.__particles.items()
+
+    def keys(self) -> KeysView[str]:
+        return self.__particles.keys()
+
+    def values(self) -> ValuesView[Particle]:
+        return self.__particles.values()
```

Comment:
I think this should inherit from collections base class, as the behavior is something like a dict but not completely. Which could be confusing to a user.

Reply:
Actually I was searching for a `dict` base class, but there is no such `abc` as far as I could find, also not in the `typing` library. And technically, this `ParticleCollection` is like a 'read-only `dict`'.

Reply:
So one example I'm referring to is the `__iter__` method. I guess this should be a `Mapping` + the add function

Reply:
Yes, this is a good point actually... The missing `__iter__` method actually causes a subtle bug in the tests, which, clumsily, had been avoided by iterating over `.keys()`.

Reply:
So you say we would need to define our own abstract base class? I cannot think of a built-in abc that has this kind of read-only interface.

Reply:
No, just use the matching collection abc. So `class ParticleCollection(abc.Mapping):`. Then we ensure it behaves just like a Mapping. In addition we add this `add` method (lots of adds :smile: ).

Reply:
Maybe this also suggests that we are using a wrong thing. Theoretically you don't need the keys and you just have a hash set based on the pid or name.

Reply:
Yeah sorry I just realised when looking at the `collections.abc` class... :see_no_evil:

Reply:
> Maybe this also suggests that we are using a wrong thing. Theoretically you don't need the keys and you just have a hash set based on the pid or name.

I don't think so. One needs some sort of lookup, right?

Reply:
True